### PR TITLE
liveness probe for pithos container

### PR DESCRIPTION
**Pull request by UnderGreen**

### CODE REVIEW

File: images/healthz/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,6 @@
+FROM quay.io/gravitational/debian-tall:0.0.1
+
+ADD bin/healthz \
```
Comment (a-palchikov) [inline images/healthz/Dockerfile]: `ADD bin/heathz /usr/local/bin/`
it's short enough to fit on a line

File: images/cassandra/rootfs/initialize.sh [hunk 2]
```diff
@@ -9,4 +9,6 @@ fi
 
 sed -i 's/REPLICATION_FACTOR/'"${REPLICATION_FACTOR}"'/g' /init.cql
 
-cqlsh -f /init.cql cassandra.default.svc.cluster.local
+if ! cqlsh cassandra.default.svc.cluster.local -k storage -e "select * from storage.block"; then
```
Comment (a-palchikov) [inline images/cassandra/rootfs/initialize.sh]: is this to ensure that `init.cql` is not run on a database where the select returns something?
can it return an empty dataset?
can the cql script be written so it creates or updates?
The reason is that basing this decision on `cqlsh` failing or succeeding is a bit dangerous - it does not have enough context to make a decision about whether to initialize (or does it?)

File: tool/healthz/Makefile [hunk 3]
```diff
@@ -0,0 +1,18 @@
+.PHONY: all
+all: bin/healthz
+
+NOROOT=-u $$(id -u):$$(id -g)
+SRCDIR=/go/src/github.com/gravitational/pithos-app
+DOCKERFLAGS=--rm=true $(NOROOT) -v $(PWD):$(SRCDIR) -w $(SRCDIR) -e GO15VENDOREXPERIMENT=1
```
Comment (a-palchikov) [inline tool/healthz/Makefile]: ~~GO15VENDOREXPERIMENT=1~~

File: tool/healthz/main.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+)
+
+const (
+	defaultEndpoint       = "localhost:18080"
+	defaultBucket         = "liveness-check"
+	defaultBucketLocation = "G1"
+	defaultPrefix         = "liveness"
+)
+
+// Config for access to S3 storage
```
Comment (a-palchikov) [inline tool/healthz/main.go]: `s3Config describes configuration to access an s3 bucket`

File: tool/healthz/main.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+type s3Config struct {
+	bucket string
+	client *minio.Client
+}
+
+func main() {
+	s3AccessKeyID := flag.String("access-key", "", "S3 access key")
```
Comment (a-palchikov) [inline tool/healthz/main.go]: For compatibility with `AWS`, I'd call these `access key id` and `secret access key`

File: tool/healthz/main.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	s3Endpoint := flag.String("endpoint", defaultEndpoint, "S3 endpoint address")
+	s3Bucket := flag.String("bucket", defaultBucket, "S3 Bucket name")
+
+	flag.Parse()
+	log.Infof("starting healthz endpoint")
+
+	http.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
+		if err := livenessProbe(*s3Endpoint, *s3Bucket, *s3AccessKeyID, *s3SecretAccessKey); err != nil {
+			log.Error(err)
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (a-palchikov) [inline tool/healthz/main.go]: `ServiceUnavailable` (`503`) feels more appropriate for a liveness probe failure

File: tool/healthz/main.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+	log.Fatal(http.ListenAndServe(":8080", nil))
+}
+
+func livenessProbe(s3Endpoint, s3Bucket, s3AccessKeyID, s3SecretAccessKey string) error {
+	var err error
+	var client *minio.Client
+
+	if s3AccessKeyID == "" && s3SecretAccessKey == "" {
+		log.Fatal("s3AccessKey and s3SecretAccessKey is undefined")
+		os.Exit(1)
```
Comment (a-palchikov) [inline tool/healthz/main.go]: Should not do this. Return error and verify the keyID/secret prior to executing this function.

File: tool/healthz/main.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s3Config := &s3Config{
+		bucket: s3Bucket,
+		client: client,
+	}
+
+	log.Info("checking creating of S3 bucket")
```
Comment (a-palchikov) [inline tool/healthz/main.go]: `verify that can create S3 bucket`
`verify that can create an object`
Do you need these messages? Is the log going to contain enough context if any of these fail?

File: tool/healthz/main.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	log.Info("teardown bucket and objects")
+	if err := s3Config.deleteBucket(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func initClient(endpoint, accessKeyID, secretAccessKey string) (*minio.Client, error) {
+	client, err := minio.NewV2(endpoint, accessKeyID, secretAccessKey, false)
```
Comment (a-palchikov) [inline tool/healthz/main.go]: can you assign `false` to a variable that bears a meaning:
```go
insecure := false
client, err := minio.NewV2(endpoint, accessKeyID, secretAccessKey, insecure)
```

File: tool/healthz/main.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+		}
+	}
+	return nil
+}
+
+func (s3c *s3Config) createObject() error {
+	var content = []byte("test")
+	reader := bytes.NewReader(content)
+
+	t := time.Now()
```
Comment (a-palchikov) [inline tool/healthz/main.go]:
```go
now := time.Now()
```
is more readable

File: tool/healthz/main.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+	_, err := s3c.client.PutObject(s3c.bucket, objectName, reader, "application/octet-stream")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (s3c *s3Config) deleteBucket() error {
+	// Create a done channel to control 'ListObjectsV2' go routine.
```
Comment (a-palchikov) [inline tool/healthz/main.go]: `goroutine`

File: tool/healthz/main.go [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+	doneCh := make(chan struct{})
+
+	// Indicate to our routine to exit cleanly upon return.
```
Comment (a-palchikov) [inline tool/healthz/main.go]: Unhelpful comment.
If you meant closing a channel as `a clean exit`, then say so.

File: tool/healthz/main.go [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+	defer close(doneCh)
+
+	isRecursive := false
```
Comment (a-palchikov) [inline tool/healthz/main.go]: i mean, `recursive` is already an adjective

File: images/cassandra/rootfs/initialize.sh (hunk 2 above)
Comment (federicoboerr) [reply images/cassandra/rootfs/initialize.sh]: Agree with @a-palchikov. I debugged this issue last Friday and wrote an extent summary of it at AOP-1609.
The bottom line is:
```
There are 3 things to do here:
1. Make sure cassandra is responding correctly before executing the command
2. Increate cql command timeout
3. Make the cql script or the shell script to not fail if the assets exist
```

Comment (federicoboerr) [reply images/cassandra/rootfs/initialize.sh]: The real solution is to make sure that `init.cql` is idempotent.

Comment (federicoboerr) [reply images/cassandra/rootfs/initialize.sh]: `CREATE TABLE IF NOT EXISTS keyspace_name.table_name` can be the solution for every statement.

File: resources/pithos.yaml [hunk 14]
```diff
@@ -121,6 +121,12 @@ spec:
             mountPath: /etc/cluster-ssl
       - image: pithos-healthz:latest
         name: healthz
+        args:
```
Comment (federicoboerr) [inline resources/pithos.yaml]: Why not put this as the `CMD` of the docker image?

**Comment by UnderGreen** [comment]
Closing this prior to creating new smaller PR's.

**Pull request closed without merging**